### ISSUE TITLE
ENH: add isotonic regression

### ISSUE THREAD

**Issue by lorentzenchr** [scipy/scipy#17706]
### Is your feature request related to a problem? Please describe.

According to Busing 2022, nonparametric isotonic (or monotone) regression
> [...] has applications in fields like operations research, signal processing, and statistics, (see, e.g., Kyng, Rao, and Sachdeva 2015; Chatterjee, Guntuboyina, and Sen 2015). More specifically, concerning statistics, the pool-adjacent-violators algorithm and its implementations have applications in, for example, test statistics (Bartholomew 1961), maxi- mum likelihood estimation of ordered means (Robertson 1978), unimodal function estimation (Turner and Wollan 1997; Eggermont and LaRiccia 2000; Reboul 2005), calculation of ad- justed response probabilities (Pace, Stylianou, and Warltier 2007), assessing the quality of micro-array titration data (Klinglmueller 2010), and efficient first-order generalized gradient algorithms (Tibshirani and Suo 2016). In our case, monotone regression is used as part of a larger procedure, like for one of its originators (Kruskal 1964a,b): the transformation step in a nonmetric multidimensional scaling algorithm.

On top, it is used in reliability diagrams and calibration of classifiers.

**References**
- Busing, F. M. T. A. (2022). Monotone Regression: A Simple and Fast O(n) PAVA Implementation. Journal of Statistical Software, Code Snippets, 102(1), 1–25. https://doi.org/10.18637/jss.v102.c01
  R package: https://cran.r-project.org/package=monotone

### Describe the solution you'd like.

Implement isotonic regression via the O(n) pool-adjacent-violaters (PAV) algorithm, Algorithm 1 of Busing 2022.

### Describe alternatives you've considered.

Use scikit-learns implementation, https://scikit-learn.org/stable/modules/generated/sklearn.isotonic.isotonic_regression.html#sklearn-isotonic-isotonic-regression.

2 points in favor of having it in scipy:
- I guess, if it were available in scipy, scikit-learn would be happy to use scipy's implementation under the hood.
- I would not search in scikit-learn for a univariate non-parametric regression method and think that `scipy.interpolate` would be a very good place for it.

**Comment by anntzer** [comment]
FWIW I wrote most of the main isotonic regression loop in scikit-learn (_inplace_contiguous_isotonic_regression, https://github.com/scikit-learn/scikit-learn/pull/6691 / https://github.com/scikit-learn/scikit-learn/pull/7444); I would be happy if that got lifted into scipy itself.

**Comment by ev-br** [comment]
Can you explain a bit what is this isotonic regression? 
I seem to remember bug reports with the word "isotonic" talking about `interp1d(..., kind="slinear")`, but that's all.
The CRAN link talks about "monotone interpolants", but these seem to be different from PCHIP and Akima1D monotone C1 cubics that scipy interpolate provides.

Scikit-learn user guide, https://scikit-learn.org/stable/modules/isotonic.html#isotonic leaves me wondering what is the class of functions it works in  --- "fits a non-decreasing real function to 1-dimensional data" --- what exactly is this function. From the plots on that page I'm also not sure if it's a piecewise constant or piecewise linear or something else?

**Comment by lorentzenchr** [comment]
In short, isotonic regression solves $$\operatorname{argmin}_{x_i} \sum_i (y_i - x_i)^2 \quad\text{subject to } x_i \leq x_j \text{ whenever } i \leq j$$

This is a monotone increasing interpolation of the data points $y$: pools/blocks/regions of constant values (the adjacent pool violations) and then a freedom to interpolate between such regions, often linearly.
Often one has a covariate, say z, sorts it and then sort y accordingly (with the same indices of argsort). Then one uses isotonic regression in the sorted y which in total is isotonic regression of y vs z. The scikit-learn user guide demonstrates its use, see https://scikit-learn.org/stable/modules/isotonic.html#isotonic-regression.
![image](https://user-images.githubusercontent.com/15324633/210559526-7a61ccec-41ac-4dd6-8761-8c6994a90972.png)

Note that the squared loss can be replaced by any Bregman function (loss/scoring functions that are consistent for the mean functional), the isotonic regression result stays the same! Further details of isotonic regression and links to literature are given in

Jordan, A.I., Mühlemann, A. & Ziegel, J.F. Characterizing the optimal solutions to the isotonic regression problem for identifiable functionals. Ann Inst Stat Math 74, 489–514 (2022). https://doi.org/10.1007/s10463-021-00808-0 ([link to pdf](https://www.ism.ac.jp/editsec/aism/74/0489.pdf))

**Comment by lorentzenchr** [comment]
For info: I'm preparing a PR.

**Comment by ev-br** [comment]
Repeating the scikit-learn user guide does not help really :-). 

The formula at the top does not define an interpolant --- it very well might be that I'm just being dense, so please bear with me. 

Could you rephrase it in the following language (will need it anyway if this is to be in scipy.interpolate). Given two arrays, `xi` ("data points") and `yi` ("values") of length `N` each, does this isotonic thing interpolates the data or approximates the data?

What I mean is: an interpolant defines a certain function, `f(x)` which interpolates the data: 

`f(xi[j]) = yi[j]` for `j=0, ..., N-1`

If it's an approximation not interpolation, there is still some function `f(x; p)`, of data points `x` and parameters `p`. You then select the parameters to minimize a certain loss function with respect to `p`, e.g.

`S(p) = sum( (f(xi[j]; p) - y[j])**2 for j in range(N))`

So, for this isotonic thing, what is this function `f(x;p)`? 
Then, what are the parameters `p`?
And what is the desired smoothness of `f(x; p_optimal)`?

In this language, for the cubic spline interpolant the function `f(x; p)` is a piecewise cubic with breakpoints at the data points `xi`, the parameters are coefficients of this cubic parabolas, which you select to have a C2 smoothness. For monotone interpolants, the parameters are still coefficients of cubics, but you select them in a different way to only require C1 smoothness etc.

**Comment by WarrenWeckesser** [comment]
@ev-br, given { $(x_i, y_i)$ }, the first part of isotonic regression does not involve { $x_i$ }.  A set { $\hat{y}_i$ } is found that minimizes the squared errors between { $\hat{y}_i$ } and { $y_i$ } subject to the constraint that { $\hat{y}_i$ } is monotonic.  See the section ["Problem Statement and Algoritms"](https://en.wikipedia.org/wiki/Isotonic_regression#Problem_Statement_and_Algorithms) in the [wikipedia article](https://en.wikipedia.org/wiki/Isotonic_regression).  Once { $\hat{y}_i$ } is found, in principle any montonic interpolator could be used to connect the points in { $(x_i, \hat{y}_i)$ }.  That is the gist of the comment that begins "To complete the isotonic regression task..." in the wikipedia article.  As noted there, a common choice is linear interpolation.

Here's a quick demo based on `minimize` (so certainly not the most efficient implementation).  The code doesn't actually create an interpolator, but the green line shows the graph of what the interpolator would be if linear interpolation is used between the points { $(x_i, \hat{y}_i)$ }.

```
import numpy as np
from scipy.optimize import minimize
import matplotlib.pyplot as plt


data = np.array([[1.2, 1.5],
                 [2.0, 1.0],
                 [3.0, 4.0],
                 [5.0, 6.0],
                 [5.5, 5.7],
                 [6.5, 5.0],
                 [7.5, 7.8],
                 [8.0, 9.0],
                 [8.5, 7.5],
                 [9.0, 9.5],
                 [9.5, 9.0]])

x, y = data.T


def objective(yhat, y):
    return np.sum((yhat - y)**2)


def constraint(yhat, y):
    # This is for a monotonically increasing regression.
    return np.diff(yhat)


result = minimize(objective, x0=y, args=(y,),
                  constraints=[{'type': 'ineq',
                                'fun': lambda x: constraint(x, y)}])

print(result)
yhat = result.x

plt.plot(x, y, 'k.', alpha=0.75, label='data')
plt.plot(x, yhat, 'go-', alpha=0.4, label='isotonic + linear interp')
plt.grid()
plt.xlabel('x')
plt.ylabel('y')
plt.legend(framealpha=1, shadow=True)

plt.show()
```
That script generates this plot:

![figure](https://user-images.githubusercontent.com/321463/210578510-3a058239-9abd-4d9a-bbf6-69d498683b95.png)

**Comment by ev-br** [comment]
Thanks a lot @WarrenWeckesser , this makes it way clearer! I was confused indeed. 
Would you weight in on whether this is a good fit for scipy in some form?

**Comment by lorentzenchr** [comment]
Mailing list thread: https://mail.python.org/archives/list/scipy-dev@python.org/message/ERRX6ZZAX6NROR6GEE6Q4ZA6WCIOB2ZL/

**Comment by lorentzenchr** [comment]
Hi
Can someone confirm that the message(s) on the mailing list got through? I have got no feedback at all on those.
And I currently am unavailable for community meetings to ask.

**Comment by ev-br** [comment]
Yes, it did go through. The lack of response we interpret as lazy consensus to proceed with the PR. Am going to review it, time permitting; Other reviews are most welcome! Reviews/feedback from domain experts and/or users would be superb.

**Comment by lorentzenchr** [comment]
> does this isotonic thing interpolates the data or approximates the data?

To answer explicitly within the given d definitions, isotonic regression _approximates_ the data.

> The lack of response we interpret as lazy consensus

@ev-br Thanks for clarifying.
